### PR TITLE
Render the / path of a branched resource

### PR DESCRIPTION
**Pull request by notoriousno**
Address #237

### CODE REVIEW

**Comment by glyph** [comment]
It feels like this should be a redirect since the trailing slash included should be the canonical form of the URL.  Or there should be some other annotation allowing you to distinguish (mainly so that any generated relative links are correct).

More importantly though: can you appease travis? :)

**Comment by notoriousno** [comment]
> It feels like this should be a redirect since the trailing slash included should be the canonical form of the URL


Maybe but Klein doesn't "respect" the `/` path in other circumstances and `@route('')` is prohibited by `werkzueg`. For instance the following will work for `https://example.com/` or `https://example.com`:


``` python
@route('/')
def root(req):
    #...
```

I'm not opposed to the idea of some formal way of representing "this path is the root resource" but I certainly cannot think of a way that isn't awkward. However, I'm more reluctant to go down the redirect route. Let's talk this one through a bit more.


> More importantly though: can you appease travis?


The error is due to `_resource.KleinResource.render` being too complex (which it is but not because of what this PR adds). This PR only adds a 2 line conditional to that function. There should be another ticket to address the complexity and should not be addressed in this PR.

**Comment by glyph** [comment]
@notoriousno FYI: I have a super limited amount of time to look at PRs, and unless CI is passing, it doesn't even make it onto my list.  Even if you have to `# noqa` the line, or ignore the "too complex" warning entirely, getting CI green will make it substantially more likely I'll catch it in the next pass :).

**Comment by notoriousno** [comment]
@glyph I'd love to `# noqa` the complex function and call it a day but it doesn't seem to be possible for some reason. I've tried putting `# noqa: C901` before, after, and on the line to no avail. Looking at `tox.ini` I see that code complexity is set to `21` which I'm sure was specifically set for `_resource.KleinResource.render`. So here are the options:

1. Increase code complexity in `tox.ini`
1. Separate out the callback functions in `render()` into their own class functions.
1. Find another less complex function where the fix can be applied.

**Comment by glyph** [comment]
I don't even know why we have a code complexity checker; I've never seen it fail in a useful way, so IMHO it's just wasting our time; the limit seems totally arbitrary and is not exerting any positive pressure on the code.

Please feel free to raise the limit to an arbitrary maximum to get this to pass (ideally in a separate PR; let's do small changes) and then maybe separately let's talk about getting rid of mccabe entirely.

**Comment by notoriousno** [comment]
Did something stall during the code coverage step?

**Comment by wsanchez** [comment]
Restarted Travis build in hopes that kicks the status check

**Comment by glyph** [comment]
oh, codecov

**Comment by glyph** [comment]
OK!

Sorry that I didn't answer the questions about the feature earlier.

Let's talk about redirects.

If you have a branched resource like this, in `/foo`, the implication is that the resource is a *collection*, since it contains other, sub-resources, in `/foo/bar`, `/foo/baz`, etc.

As such, Klein routers are always effectively collections, and can always exist, be "mounted", at the "top level" of a website.

If you have a resource at `/`, it can render links to other resources.  Relative links will be interpreted to be other resources at `/x`, `/y`, etc.

If we were to move this to `/foo`, it should behave the same, and that's only possible if `/foo` cannot be rendered.  There is no place on a web site "above" the top-level `/` URL.

> Maybe but Klein doesn't "respect" the / path in other circumstances and @route('') is prohibited by werkzueg. For instance the following will work for https://example.com/ or https://example.com:

One nuance you're missing is that `https://example.com/` and `https://example.com` are *the same URL*.  Let's ask our buddy `hyperlink`:

```
>>> from hyperlink import parse
>>> a = parse("https://example.com/")
>>> b = parse("https://example.com")
>>> a == b
True
```

Now, this is a little confusing because if we ask `hyperlink` some *other* questions, it appears justifiably confused:

```
>>> a.path
('',)
>>> b.path
()
```

but this peculiar distinction only exists for the library to have byte-identical round-tripping of the input text.  As soon as you canonicalize the representation, you'll see that they're the same:

```
>>> b.normalize()
DecodedURL(url=URL.from_text('https://example.com/'))
```

More to the point: although you can, in human-readable text, distinguish these URLs, it is not physically possible for a web server to do so, since the wire-level request generated for both is identical.  Here, let's ask [`httpie`](https://httpie.org):

```console
$ http -p HB https://example.com/
GET / HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: example.com
User-Agent: HTTPie/0.9.9

$ http -p HB https://example.com
GET / HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: example.com
User-Agent: HTTPie/0.9.9
```

File: src/klein/test/test_resource.py [hunk 1]
```diff
@@ -1167,6 +1167,29 @@ def real(req):
...
 
+    def test_branchRoot(self):
+        """
+        Return the '/' path of a branched resource
+        """
+        expected_result = b'foo'
+        app = self.app
+        branch = Klein()
+
+        @app.route('/foo')
```
Comment (wsanchez) [inline src/klein/test/test_resource.py]: Should we also test what happens if you do `@app.route('/foo/')`?

**Comment by glyph** [comment]
Since it seems like the behavior implemented here is wrong, and the OP has not commented since August, I'm going to close this now to clean up `/pulls`.  But please do feel free to reopen if you want to work on it some more!

**Pull request closed without merging**